man/man8/radmin.8
'\" t
.\"     Title: radmin
.\"    Author: Alan DeKok
.\" Generator: Asciidoctor 2.0.10
.\"      Date: 2020-12-25
.\"    Manual: FreeRADIUS
.\"    Source: FreeRADIUS
.\"  Language: English
.\"
.TH "RADMIN" "8" "2020-12-25" "FreeRADIUS" "FreeRADIUS"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
radmin \- FreeRADIUS Administration tool
.SH "SYNOPSIS"
.sp
\fBradmin\fP [\fB\-d\fP \fIconfig_directory\fP] [\fB\-e\fP \fIcommand\fP] [\fB\-E\fP] [\fB\-f\fP
\fIsocket_file\fP] [\fB\-h\fP] [\fB\-i\fP \fIinput_file\fP] [\fB\-l\fP \fIlog_file\fP] [\fB\-n\fP
\fIname\fP] [\fB\-q\fP]
.SH "DESCRIPTION"
.sp
FreeRADIUS Server administration tool that connects to the control
socket of a running server, and gives a command\-line interface to it.
.sp
At this time, only a few commands are supported. Please type \f(CRhelp\fP at
the command prompt for detailed information about the supported
commands.
.sp
The \f(CRradmin\fP command supports full tab completion, inline help via \f(CR?\fP
and \f(CRhelp\fP, and command history.
.SH "WARNING"
.sp
The security protections offered by this command are limited to the
permissions on the Unix domain socket, and the server configuration. If
someone can connect to the Unix domain socket, they have a substantial
amount of control over the server.
.SH "OPTIONS"
.sp
The following command\-line options are accepted by the program.
.sp
\fB\-d config_directory\fP
.RS 4
Defaults to \fI/etc/raddb\fP. \fBradmin\fP looks here for the server
configuration files to find the "listen" section that defines the
control socket filename.
.RE
.sp
\fB\-e command\fP
.RS 4
Run \fIcommand\fP and exit.
.RE
.sp
\fB\-E\fP
.RS 4
Echo commands as they are being executed.
.RE
.sp
\fB\-f socket_file\fP
.RS 4
Specify the socket filename directly. The radiusd.conf file is not
read.
.RE
.sp
\fB\-h\fP
.RS 4
Print usage help information.
.RE
.sp
\fB\-i input_file\fP
.RS 4
Reads input from the specified file. If this option is not
specified, \f(CRstdin\fP is used.  This also sets \f(CR\-q\fP.
.RE
.sp
\fB\-l log_file\fP
.RS 4
Writes the commands which are executed to this log file. This
functionality is off by default.
.RE
.sp
\fB\-n name\fP
.RS 4
Read \f(CRraddb/name.conf\fP instead of \f(CRraddb/radiusd.conf\fP.
.RE
.sp
\fB\-q\fP
.RS 4
Quiet mode.
.RE
.SH "COMMANDS"
.sp
The commands implemented by the command\-line interface are almost
completely controlled by the server. There are a few commands
interpreted locally by radmin:
.sp
\fBexit\fP
.RS 4
Exit from radmin.
.RE
.sp
\fBquit\fP
.RS 4
Exit from radmin.
.RE
.sp
\fBreconnect\fP
.RS 4
Reconnect to the server.
.RE
.sp
The other commands are implemented by the server. Type \f(CRhelp\fP at the
prompt for more information.
.SH "EXAMPLES"
.sp
\fBdebug file /var/log/radius/bob.log\fP
.RS 4
Set debug logs to \f(CR/var/log/radius/bob.log\fP. There is very little
checking of this filename. Rogue administrators may be able use this
command to over\-write almost any file on the system.  However, if
the rogue administrators have write access to the main \f(CRradius.conf\fP
file, then they can do the same thing without \f(CRradmin\fP, too.
.RE
.sp
\fBdebug condition \(aq(User\-Name == "bob")\(aq\fP
.RS 4
Enable debugging output for all requests that match the condition. Any
\f(CRunlang\fP condition is valid here. The condition is parsed as a string,
so it must be enclosed in single or double quotes. Strings enclosed in
double\-quotes must have back\-slashes and the quotation marks escaped
inside of the string.
.sp
Only one debug condition can be active at a time.
.RE
.sp
\fBdebug condition \(aqUser\-Name == ""bob"") || (Packet\-Src\-IP\-Address == 192.0.2.22\(aq\fP
.RS 4
A more complex condition that enables debugging output for requests
containing User\-Name "bob", or requests that originate from source IP
address 192.0.2.22.
.RE
.sp
\fBdebug condition\fP
.RS 4
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.  sp -1
.  IP \(bu 2.3
.\}
Disable debug conditionals.
.RE
.RE
.SH "FULL LIST OF COMMANDS"
.sp
Connect to the server and type \f(CRhelp\fP for a full list of commands.
.SH "SEE ALSO"
.sp
unlang(5), radiusd.conf(5), raddb/sites\-available/control\-socket
.SH "AUTHOR"
.sp
Alan DeKok <\c
.MTO "aland\(atfreeradius.org" "" ">"
.SH "AUTHOR"
.sp
Alan DeKok

doc/ChangeLog
FreeRADIUS 3.0.22 Tue 24 Mar 2020 12:00:00 EDT urgency=low
	Feature improvements
	* Many new "unlang" documentation files.  See "make docsite"
	  and doc/antora
	* Update dictionary.bskyb.  Fixes #3339.
	* Update alcatel.sr. #3495
	* Abfab-tls updates from Alejandro Perez.
	* Allow CoA and Disconnect messages over TLS sockets.
	  Fixes #3356.
	* New xlat for setting status of rlm_always instances and new
	  resource-check example virtual server for manipulating control flow
	  in unlang policies based on status of some external resource.
	  Patches from Terry Burton.
	* Allow delta CRLs.  Fixes #3393.
	* Limited support for dynamic home servers.  See proxy.conf
	  and doc/configuration/dynamic_home_servers.md
	* Update radmin to show more information about the home servers
	  using "show home_server list all".
	* The default configuration now replies with EAP-Key-Name, if it
	  is available, and was requested.
	* Include extensions in generated certificates.
	* Ignore user-provided dhparams in FIPS mode.
	  Patch from Alexander Scheel.
	* Remove native support for Cisco LEAP.  It is insecure, and
	  should not be used.  Proxying LEAP is still supported.
	* Automatically set fragment size / MTU, so that PEAP/EAP-TLS
	  works, and no longer requires manual changes to the configuration.
	* Allow use of password preparation methods with rlm_eap_pwd.
	  Patch from Michael Braun.
	* Add TLS 1.2 support for EAP-Fast.  Patches from Alex Clouter.
	* Add a_path_reload_interval option for tls.  See mods-available/eap.
	  Fixes #1831.  Patch from Boris Lytochkin.
	* Many, many, improvements for DHCP from Nick Porter and Terry Burton.
	* Use DHCP-specific schemas.
	* Add stored procedures for DHCP lease allocation
	* Add support for DHCP-Decline.
	* More RFC complaince for various corner cases of DHCP,
	* Treat DHCP Discover and Request differently for lease allocation times.
	* Add support for PBKDF2 keys. Fixes #2649
	* Update dictionary.mikrotik, dictionary.aruba, dictionary.paloalto, dictionary.juniper.
	* Update default PostgreSQL schema to use "text" instead of
	  a fixed-size "varchar".
	* The default minimum TLS version is not TLS 1.2
	  Older versions can be allowed by setting tls_min_version
	* Add "tls_min_version" to ldap module configuration.
	* Added mods-available/dhcp_sql which is a DHCP-specific instance
	  of the SQL module.
	* Add radmin command "show client list verbose", which gives a lot
	  more information about each client.
	* Add support for EAPS-AKA authentication to rlm_wimax.
	  Fixes #3801. Patches from Benjamin Thompson.
	* Add rlm_rest support for HTTP/2.
	* Add REST-HTTP-Status-Code attribute holding HTTP status code.
	* Add option to set http_negotiation in rlm_rest.  Fixes #2821.

	Bug fixes
	* Respect the "log_reject" configuration item in more places.
	  Fixes #3352.  This lowers the number of "Login incorrect"
	  messages when "log_reject = no".
	* Fix rpmbuild for Centos > 6.  Patch from Matthew Newton.
	* Run Post-Proxy-Type Fail... when all home servers are down.
	* Note that rlm_replicate can only use UDP, and not TCP or TLS.
	* DHCP pool lookup is now keyed by Client Identifier (Option 61) when
	  supplied by client, otherwise the hardware address is used. Compliant
	  with RFC 2132. This change will not affect existing systems on upgrade,
	  but new installations will use the new behavior.
	  Patch from Terry Burton.
	* Fix minor spelling mistakes in man pages.  Patch from
	  Alexander Scheel.
	* Don't print invalid tags in rlm_cache, among other places.
	  Fixes #3491
	* Do home_server failover immediately when an initial TCP / Radsec
	  connection fails.
	* Port EAP-PWD constant time fixes from "master" branch.  The issue
	  was verified by Mohamed Sabt, and a patch supplied by
	  Daniel De Almeida Braga.
	* Clear error on SQLITE_BUSY to prevent memory leak in corner cases.
	  Patch from Nick Porter.
	* Properly add SQL clients to virtual servers.
	* Update documentation for cert generation.  Patch from Alexander Scheel.
	* Use better API when decoding DHCP packets, to avoid unnecessary work.
	  This improves performance noticably.
	* Parse locale-dependent dates.  Fixes #3602.
	* Strip out "-frecord-gcc-switches" from rlm_python3 configure build. Fixes #3693
	* Fix radiusd.conf ENV LD_PRELOAD function.
	* Update the "sql" module so that it uses fewer handles for group selection,
	  which means that it is less likely to complain that the
	  connection pool is exhausted.
	* Update the "sql" module to return "ok" when no rows have been updated
	  for accounting on/off.
	* Make the "date" module handle UTC more consistently.
	* Check for, and complain about, inconsistent use of tls_min_version
	  versus disable_tlsv1
	* Fix "read client from SQL" code so that it properly ties clients
	  to a virtual server.  Also document the behavior.
	* Update / correct data types in dictionary.wimax
	* Fix edge case in rlm_rest post decoder which could lead to the value of
	  a post attribute being lost in the case where the output buffer was completely
	  full after writing an attribute value, and more attributes needed to be encoded.
	  Reported by Adrian Smith

FreeRADIUS 3.0.21 Tue 24 Mar 2020 12:00:00 EDT urgency=low
	Feature improvements
	* New stored procedure for allocating IPs with PostgreSQL.
	  Rates of 1500 IPs per second are now possible.
	  See raddb/mods-config/sql/ippool/postgresql/procedure.sql
	  Patch from Terry Burton.
	* Add SQL IP pool support for Microsoft SQL Server
	  See raddb/mods-config/sql/ippool/mssql/
	  Patch from Terry Burton.
	* Added RCNTEC dictionary. Closes #3168.
	* Added Pica8 dictionary. Closes #3179.
	* Add TLS-Client-Cert-Valid-Since attribute holding notBefore date.
	  Patch from Boris Lytochkin.  Fixes #3157.
	* Generate attributes containing unknown OIDs.
	  See raddb/sites-available/tls.  Patch from Boris Lytochkin.
	* Update the WiMAX dictionary.
	* Added ability to rlm_python(Python2) show a stacktrace from errors. #2979

src/main/radmin.c
/*
 * radmin.c	RADIUS Administration tool.
 *
 * Version:	$Id$
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 * Copyright 2012   The FreeRADIUS server project
 * Copyright 2012   Alan DeKok <aland@deployingradius.com>
 */

RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/md5.h>
#include <freeradius-devel/channel.h>

#include <pwd.h>
#include <grp.h>

#ifdef HAVE_GETOPT_H
#  include <getopt.h>
#endif

#ifdef HAVE_SYS_STAT_H
#  include <sys/stat.h>
#endif

#ifndef READLINE_MAX_HISTORY_LINES
#	define READLINE_MAX_HISTORY_LINES 1000
#endif

#ifdef HAVE_LIBREADLINE

# include <stdio.h>
#if defined(HAVE_READLINE_READLINE_H)
#  include <readline/readline.h>
#  define USE_READLINE (1)
#elif defined(HAVE_READLINE_H)
#  include <readline.h>
#  define USE_READLINE (1)
#endif /* !defined(HAVE_READLINE_H) */

#ifdef HAVE_READLINE_HISTORY
#  if defined(HAVE_READLINE_HISTORY_H)
#    include <readline/history.h>
#    define USE_READLINE_HISTORY (1)
#  elif defined(HAVE_HISTORY_H)
#    include <history.h>
#    define USE_READLINE_HISTORY (1)
#endif /* defined(HAVE_READLINE_HISTORY_H) */

#endif /* HAVE_READLINE_HISTORY */

#endif /* HAVE_LIBREADLINE */

/*
 *	For configuration file stuff.
 */
static char const *progname = "radmin";
static char const *radmin_version = "radmin version " RADIUSD_VERSION_STRING
#ifdef RADIUSD_VERSION_COMMIT
" (git #" STRINGIFY(RADIUSD_VERSION_COMMIT) ")"
#endif
#ifndef ENABLE_REPRODUCIBLE_BUILDS
", built on " __DATE__ " at " __TIME__
#endif
;


/*
 *	The rest of this is because the conffile.c, etc. assume
 *	they're running inside of the server.  And we don't (yet)
 *	have a "libfreeradius-server", or "libfreeradius-util".
 */
main_config_t main_config;

static bool echo = false;
static char const *secret = "testing123";

#include <sys/wait.h>

#ifdef HAVE_PTHREAD_H
pid_t rad_fork(void)
{
	return fork();
}
pid_t rad_waitpid(pid_t pid, int *status)
{
	return waitpid(pid, status, 0);
}
#endif

static void NEVER_RETURNS usage(int status)
{
	FILE *output = status ? stderr : stdout;
	fprintf(output, "Usage: %s [ args ]\n", progname);
	fprintf(output, "  -d raddb_dir    Configuration files are in \"raddbdir/*\".\n");
	fprintf(output, "  -D <dictdir>    Set main dictionary directory (defaults to " DICTDIR ").\n");
	fprintf(output, "  -e command      Execute 'command' and then exit.\n");
	fprintf(output, "  -E              Echo commands as they are being executed.\n");
	fprintf(output, "  -f socket_file  Open socket_file directly, without reading radius.conf\n");
	fprintf(output, "  -h              Print usage help information.\n");
	fprintf(output, "  -i input_file   Read commands from 'input_file'.\n");
	fprintf(output, "  -n name         Read raddb/name.conf instead of raddb/radiusd.conf\n");
	fprintf(output, "  -q              Quiet mode.\n");

	exit(status);
}

static int client_socket(char const *server)
{
	int sockfd;
	uint16_t port;
	fr_ipaddr_t ipaddr;
	char *p, buffer[1024];

	strlcpy(buffer, server, sizeof(buffer));

	p = strchr(buffer, ':');
	if (!p) {
		port = PW_RADMIN_PORT;
	} else {
		port = atoi(p + 1);
		*p = '\0';
	}

	if (ip_hton(&ipaddr, AF_INET, buffer, false) < 0) {
		fprintf(stderr, "%s: Failed looking up host %s: %s\n",
			progname, buffer, fr_syserror(errno));
		exit(1);
	}

	sockfd = fr_socket_client_tcp(NULL, &ipaddr, port, false);
	if (sockfd < 0) {
		fprintf(stderr, "%s: Failed opening socket %s: %s\n",
			progname, server, fr_syserror(errno));
		exit(1);
	}

	return sockfd;
}

static ssize_t do_challenge(int sockfd)
{
	ssize_t r;
	fr_channel_type_t channel;
	uint8_t challenge[16];

	challenge[0] = 0;

	/*
	 *	When connecting over a socket, the server challenges us.
	 */
	r = fr_channel_read(sockfd, &channel, challenge, sizeof(challenge));
	if (r <= 0) return r;

	if ((r != 16) || (channel != FR_CHANNEL_AUTH_CHALLENGE)) {
		fprintf(stderr, "%s: Failed to read challenge.\n",
			progname);
		exit(1);
	}

	fr_hmac_md5(challenge, (uint8_t const *) secret, strlen(secret),
		    challenge, sizeof(challenge));

	r = fr_channel_write(sockfd, FR_CHANNEL_AUTH_RESPONSE, challenge, sizeof(challenge));
	if (r <= 0) return r;

	/*
	 *	If the server doesn't like us, he just closes the
	 *	socket.  So we don't look for an ACK.
	 */

	return r;
}


/*
 *	Returns -1 on error.  0 on connection failed.  +1 on OK.
 */
static ssize_t run_command(int sockfd, char const *command,
			   char *buffer, size_t bufsize)
{
	ssize_t r;
	uint32_t status;
	fr_channel_type_t channel;

	if (echo) {
		fprintf(stdout, "%s\n", command);
	}

	/*
	 *	Write the text to the socket.
	 */
	r = fr_channel_write(sockfd, FR_CHANNEL_STDIN, command, strlen(command));
	if (r <= 0) return r;

	while (true) {
		r = fr_channel_read(sockfd, &channel, buffer, bufsize - 1);
		if (r <= 0) return r;

		buffer[r] = '\0';	/* for C strings */

		switch (channel) {
		case FR_CHANNEL_STDOUT:
			fprintf(stdout, "%s", buffer);
			break;

		case FR_CHANNEL_STDERR:
			fprintf(stderr, "ERROR: %s", buffer);
			break;

		case FR_CHANNEL_CMD_STATUS:
			if (r < 4) return 1;

			memcpy(&status, buffer, sizeof(status));
			status = ntohl(status);
			return status;

		default:
			fprintf(stderr, "Unexpected response\n");
			return -1;
		}
	}

	/* never gets here */
}

static int do_connect(int *out, char const *file, char const *server)
{
	int sockfd;
	ssize_t r;
	fr_channel_type_t channel;
	char buffer[65536];

	uint32_t magic;

	/*
	 *	Close stale file descriptors
	 */
	if (*out != -1) {
		close(*out);
		*out = -1;
	}

	if (file) {
		/*
		 *	FIXME: Get destination from command line, if possible?
		 */
		sockfd = fr_socket_client_unix(file, false);
		if (sockfd < 0) {
			fr_perror("radmin");
			if (errno == ENOENT) {
					fprintf(stderr, "Perhaps you need to run the commands:");
					fprintf(stderr, "\tcd /etc/raddb\n");
					fprintf(stderr, "\tln -s sites-available/control-socket "
						"sites-enabled/control-socket\n");
					fprintf(stderr, "and then re-start the server?\n");
			}
			return -1;
		}
	} else {
		sockfd = client_socket(server);
	}

	/*
	 *	Only works for BSD, but Linux allows us
	 *	to mask SIGPIPE, so that's fine.
	 */
#ifdef SO_NOSIGPIPE
	{
		int set = 1;

		setsockopt(sockfd, SOL_SOCKET, SO_NOSIGPIPE, (void *)&set, sizeof(int));
	}
#endif

	/*
	 *	Set up the initial header data.
	 */
	magic = 0xf7eead16;
	magic = htonl(magic);
	memcpy(buffer, &magic, sizeof(magic));
	memset(buffer + sizeof(magic), 0, sizeof(magic));

	r = fr_channel_write(sockfd, FR_CHANNEL_INIT_ACK, buffer, 8);
	if (r <= 0) {
	do_close:
		fprintf(stderr, "%s: Error in socket: %s\n",
			progname, fr_syserror(errno));
		close(sockfd);
			return -1;
	}

	r = fr_channel_read(sockfd, &channel, buffer + 8, 8);
	if (r <= 0) goto do_close;

	if ((r != 8) || (channel != FR_CHANNEL_INIT_ACK) ||
	    (memcmp(buffer, buffer + 8, 8) != 0)) {
		fprintf(stderr, "%s: Incompatible versions\n", progname);
		close(sockfd);
		return -1;
	}

	if (server && secret) {
		r = do_challenge(sockfd);
		if (r <= 0) goto do_close;
	}

	*out = sockfd;

	return 0;
}

#define MAX_COMMANDS (4)

int main(int argc, char **argv)
{
	int		argval;
	bool		quiet = false;
	int		sockfd = -1;
	char		*line = NULL;
	ssize_t		len;
	char const	*file = NULL;
	char const	*name = "radiusd";
	char		*p, buffer[65536];
	char const	*input_file = NULL;
	FILE		*inputfp = stdin;
	char const	*server = NULL;

	char const	*radius_dir = RADIUS_DIR;
	char const	*dict_dir = DICTDIR;
#ifdef USE_READLINE_HISTORY
	char 		history_file[PATH_MAX];
#endif

	char *commands[MAX_COMMANDS];
	int num_commands = -1;

	int exit_status = EXIT_SUCCESS;

#ifndef NDEBUG
	if (fr_fault_setup(getenv("PANIC_ACTION"), argv[0]) < 0) {
		fr_perror("radmin");
		exit(EXIT_FAILURE);
	}
#endif

	talloc_set_log_stderr();

	if ((progname = strrchr(argv[0], FR_DIR_SEP)) == NULL) {
		progname = argv[0];
	} else {
		progname++;
	}

	while ((argval = getopt(argc, argv, "d:D:hi:e:Ef:n:qs:S")) != EOF) {
		switch (argval) {
		case 'd':
			if (file) {
				fprintf(stderr, "%s: -d and -f cannot be used together.\n", progname);
				exit(1);
			}
			if (server) {
				fprintf(stderr, "%s: -d and -s cannot be used together.\n", progname);
				exit(1);
			}
			radius_dir = optarg;
			break;

		case 'D':
			dict_dir = optarg;
			break;

		case 'e':
			num_commands++; /* starts at -1 */
			if (num_commands >= MAX_COMMANDS) {
				fprintf(stderr, "%s: Too many '-e'\n",
					progname);
				exit(1);
			}

			commands[num_commands] = optarg;
			break;

		case 'E':
			echo = true;
			break;

		case 'f':
			radius_dir = NULL;
			file = optarg;
			break;

		default:
		case 'h':
			usage(0);	/* never returns */

		case 'i':
#ifdef __clang_analyzer__
			if (!optarg) exit(1);
#endif

			if (strcmp(optarg, "-") != 0) {
				input_file = optarg;
			}
			quiet = true;
			break;

		case 'n':
			name = optarg;
			break;

		case 'q':
			quiet = true;
			break;

		case 's':
			if (file) {
				fprintf(stderr, "%s: -s and -f cannot be used together.\n", progname);
				usage(1);
			}
			radius_dir = NULL;
			server = optarg;
			break;

		case 'S':
			secret = NULL;
			break;
		}
	}

	/*
	 *	Mismatch between the binary and the libraries it depends on
	 */
	if (fr_check_lib_magic(RADIUSD_MAGIC_NUMBER) < 0) {
		fr_perror("radmin");
		exit(1);
	}

	if (radius_dir) {
		int rcode;
		CONF_SECTION *cs, *subcs;
		uid_t		uid;
		gid_t		gid;
		char const	*uid_name = NULL;
		char const	*gid_name = NULL;
		struct passwd	*pwd;
		struct group	*grp;

		file = NULL;	/* MUST read it from the conffile now */

		snprintf(buffer, sizeof(buffer), "%s/%s.conf", radius_dir, name);

		/*
		 *	Need to read in the dictionaries, else we may get
		 *	validation errors when we try and parse the config.
		 */
		if (dict_init(dict_dir, RADIUS_DICTIONARY) < 0) {
			fr_perror("radmin");
			exit(64);
		}

		if (dict_read(radius_dir, RADIUS_DICTIONARY) == -1) {
			fr_perror("radmin");
			exit(64);
		}

		cs = cf_section_alloc(NULL, "main", NULL);
		if (!cs) exit(1);

		if (cf_file_read(cs, buffer) < 0) {
			fprintf(stderr, "%s: Errors reading or parsing %s\n", progname, buffer);
			talloc_free(cs);
			usage(1);
		}

		uid = getuid();
		gid = getgid();

		subcs = NULL;
		while ((subcs = cf_subsection_find_next(cs, subcs, "listen")) != NULL) {
			char const *value;
			CONF_PAIR *cp = cf_pair_find(subcs, "type");

			if (!cp) continue;

			value = cf_pair_value(cp);
			if (!value) continue;

			if (strcmp(value, "control") != 0) continue;

			/*
			 *	Now find the socket name (sigh)
			 */
			rcode = cf_item_parse(subcs, "socket", FR_ITEM_POINTER(PW_TYPE_STRING, &file), NULL);
			if (rcode < 0) {
				fprintf(stderr, "%s: Failed parsing listen section 'socket'\n", progname);
				exit(1);
			}

			if (!file) {
				fprintf(stderr, "%s: No path given for socket\n", progname);
				usage(1);
			}

			/*
			 *	If we're root, just use the first one we find
			 */
			if (uid == 0) break;

			/*
			 *	Check UID and GID.
			 */
			rcode = cf_item_parse(subcs, "uid", FR_ITEM_POINTER(PW_TYPE_STRING, &uid_name), NULL);
			if (rcode < 0) {
				fprintf(stderr, "%s: Failed parsing listen section 'uid'\n", progname);
				exit(1);
			}

			if (!uid_name) break;

			pwd = getpwnam(uid_name);
			if (!pwd) {
				fprintf(stderr, "%s: Failed getting UID for user %s: %s\n", progname, uid_name, strerror(errno));
				exit(1);
			}

			if (uid != pwd->pw_uid) continue;

			rcode = cf_item_parse(subcs, "gid", FR_ITEM_POINTER(PW_TYPE_STRING, &gid_name), NULL);
			if (rcode < 0) {
				fprintf(stderr, "%s: Failed parsing listen section 'gid'\n", progname);
				exit(1);
			}

			if (!gid_name) break;

			grp = getgrnam(gid_name);
			if (!grp) {
				fprintf(stderr, "%s: Failed getting GID for group %s: %s\n", progname, gid_name, strerror(errno));
				exit(1);
			}

			if (gid != grp->gr_gid) continue;

			break;
		}

		if (!file) {
			fprintf(stderr, "%s: Could not find control socket in %s\n", progname, buffer);
			exit(1);
		}
	}

	if (input_file) {
		inputfp = fopen(input_file, "r");
		if (!inputfp) {
			fprintf(stderr, "%s: Failed opening %s: %s\n", progname, input_file, fr_syserror(errno));
			exit(1);
		}
	}

	if (!file && !server) {
		fprintf(stderr, "%s: Must use one of '-d' or '-f' or '-s'\n",
			progname);
		exit(1);
	}

	/*
	 *	Check if stdin is a TTY only if input is from stdin
	 */
	if (input_file && !quiet && !isatty(STDIN_FILENO)) quiet = true;

#ifdef USE_READLINE
	if (!quiet) {
#ifdef USE_READLINE_HISTORY
		using_history();
		stifle_history(READLINE_MAX_HISTORY_LINES);

		snprintf(history_file, sizeof(history_file), "%s/%s", getenv("HOME"), ".radmin_history");
		read_history(history_file);
#endif
		rl_bind_key('\t', rl_insert);
	}
#endif

	/*
	 *	Prevent SIGPIPEs from terminating the process
	 */
	signal(SIGPIPE, SIG_IGN);

	if (do_connect(&sockfd, file, server) < 0) exit(1);

	/*
	 *	Run one command.
	 */
	if (num_commands >= 0) {
		int i;

		for (i = 0; i <= num_commands; i++) {
			len = run_command(sockfd, commands[i], buffer, sizeof(buffer));
			if (len < 0) exit(1);

			if (len == FR_CHANNEL_FAIL) exit_status = EXIT_FAILURE;
		}
		exit(exit_status);
	}

	if (!quiet) {
		printf("%s - FreeRADIUS Server administration tool.\n", radmin_version);
		printf("Copyright (C) 2008-2019 The FreeRADIUS server project and contributors.\n");
		printf("There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A\n");
		printf("PARTICULAR PURPOSE.\n");
		printf("You may redistribute copies of FreeRADIUS under the terms of the\n");
		printf("GNU General Public License v2.\n");
	}

	/*
	 *	FIXME: Do login?
	 */

	while (1) {
		int retries;

#ifndef USE_READLINE
		if (!quiet) {
			printf("radmin> ");
			fflush(stdout);
		}
#else
		if (!quiet) {
			line = readline("radmin> ");

			if (!line) break;

			if (!*line) {
				free(line);
				continue;
			}

#ifdef USE_READLINE_HISTORY
			add_history(line);
			write_history(history_file);
#endif
		} else		/* quiet, or no readline */
#endif
		{
			line = fgets(buffer, sizeof(buffer), inputfp);
			if (!line) break;

			p = strchr(buffer, '\n');
			if (!p) {
				fprintf(stderr, "%s: Input line too long\n",
					progname);
				exit(1);
			}

			*p = '\0';

			/*
			 *	Strip off leading spaces.
			 */
			for (p = line; *p != '\0'; p++) {
				if ((p[0] == ' ') ||
				    (p[0] == '\t')) {
					line = p + 1;
					continue;
				}

				if (p[0] == '#') {
					line = NULL;
					break;
				}

				break;
			}

			/*
			 *	Comments: keep going.
			 */
			if (!line) continue;

			/*
			 *	Strip off CR / LF
			 */
			for (p = line; *p != '\0'; p++) {
				if ((p[0] == '\r') ||
				    (p[0] == '\n')) {
					p[0] = '\0';
					break;
				}
			}
		}

		if (strcmp(line, "reconnect") == 0) {
			if (do_connect(&sockfd, file, server) < 0) exit(1);
			line = NULL;
			continue;
		}

		if (strncmp(line, "secret ", 7) == 0) {
			if (!secret) {
				secret = line + 7;
				do_challenge(sockfd);
			}
			line = NULL;
			continue;
		}

		/*
		 *	Exit, done, etc.
		 */
		if ((strcmp(line, "exit") == 0) ||
		    (strcmp(line, "quit") == 0)) {
			break;
		}

		if (server && !secret) {
			fprintf(stderr, "ERROR: You must enter 'secret <SECRET>' before running any commands\n");
			line = NULL;
			continue;
		}

		retries = 0;
	retry:
		len = run_command(sockfd, line, buffer, sizeof(buffer));
		if (len < 0) {
			if (!quiet) fprintf(stderr, "... reconnecting ...\n");

			if (do_connect(&sockfd, file, server) < 0) {
				exit(1);
			}

			retries++;
			if (retries < 2) goto retry;

			fprintf(stderr, "Failed to connect to server\n");
			exit(1);

		} else if (len == FR_CHANNEL_SUCCESS) {
			break;

		} else if (len == FR_CHANNEL_FAIL) {
			exit_status = EXIT_FAILURE;
		}
	}

	fprintf(stdout, "\n");

	if (inputfp != stdin) fclose(inputfp);

	return exit_status;
}

